src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +65,740 @@ namespace Microsoft.CodeAnalysis.FindSymbols
             CancellationToken cancellationToken)
         {
-            if (type != null && type.TypeKind == TypeKind.Class)
-            {
-                return GetDependentTypesAsync(
-                    type,
-                    solution,
-                    null,
-                    (candidate, baseType) => OriginalSymbolsMatch(candidate.BaseType, baseType, solution, cancellationToken),
-                    s_derivedClassesCache,
-                    cancellationToken);
-            }
-
-            return SpecializedTasks.EmptyEnumerable<INamedTypeSymbol>();
+            return FindDerivedAndImplementingTypesAsync(
+                type, solution, projects: null,
+                transitive: false, cancellationToken: cancellationToken);
         }
 
-        public static Task<IEnumerable<INamedTypeSymbol>> GetTypesImmediatelyDerivedFromInterfacesAsync(
+        private static Task<IEnumerable<INamedTypeSymbol>> FindDerivedAndImplementingTypesAsync(
             INamedTypeSymbol type,
             Solution solution,
+            IImmutableSet<Project> projects,
+            bool transitive, 
             CancellationToken cancellationToken)
         {
-            if (type != null && type.TypeKind == TypeKind.Interface)
+            // Only an interface can be implemented.
+            if (type?.TypeKind == TypeKind.Interface)
             {
-                type = type.OriginalDefinition;
-                return GetDependentTypesAsync(
-                    type,
-                    solution,
-                    null,
-                    (candidate, baseInterface) => candidate.Interfaces.Any(i => OriginalSymbolsMatch(i, baseInterface, solution, cancellationToken)),
-                    s_derivedInterfacesCache,
-                    cancellationToken);
-            }
+                return FindTypesAsync(type, solution, projects,
+                    findTypesInProjectAsync: FindDerivedAndImplementingTypesInProjectAsync,
+                    transitive: transitive,
+                    cancellationToken: cancellationToken);
+           }
 
             return SpecializedTasks.EmptyEnumerable<INamedTypeSymbol>();
         }
 
-        private static async Task<IEnumerable<INamedTypeSymbol>> GetDependentTypesAsync(
+        private static async Task<IEnumerable<INamedTypeSymbol>> FindTypesAsync(
             INamedTypeSymbol type,
             Solution solution,
             IImmutableSet<Project> projects,
-            Func<INamedTypeSymbol, INamedTypeSymbol, bool> predicate,
-            ConditionalWeakTable<Compilation, ConcurrentDictionary<SymbolKey, List<SymbolKey>>> cache,
+            FindTypesInProjectCallback findTypesInProjectAsync,
+            bool transitive,
             CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
+            type = type.OriginalDefinition;
+            projects = projects ?? ImmutableHashSet.Create(solution.Projects.ToArray());
```

Comment:
I find it odd that null implies all.  Can/should this be done at the callers, or at least cached?